src/rgw/rgw_data_sync.cc

```diff
@@ -4136,12 +4136,79 @@ static bool has_olh_epoch(RGWModifyOp op) {
   return op == CLS_RGW_OP_LINK_OLH || op == CLS_RGW_OP_UNLINK_INSTANCE;
 }
 
+class RGWBucketShardIsDoneCR: public RGWCoroutine {
+  RGWDataSyncCtx *sc;
+  RGWDataSyncEnv *sync_env;
+  rgw_bucket_sync_status bucket_status;
+  const rgw_raw_obj status_obj;
+  const uint64_t num_shards;
+  RGWObjVersionTracker& objv_tracker;
+
+public:
+  RGWBucketShardIsDoneCR(RGWDataSyncCtx *_sc, const std::string& _status_oid,
+                         int _num_shards, RGWObjVersionTracker& _objv_tracker)
+    : RGWCoroutine(_sc->cct), sc(_sc), sync_env(_sc->env),
+      status_obj(sync_env->svc->zone->get_zone_params().log_pool, _status_oid),
+      num_shards(_num_shards), objv_tracker(_objv_tracker) {}
+
+  int operate() override
+  {
+    reenter(this) {
+      // read bucket sync status
+      using ReadCR = RGWSimpleRadosReadCR<rgw_bucket_sync_status>;
+      yield call(new ReadCR(sync_env->async_rados, sync_env->svc->sysobj,
+                            status_obj, &bucket_status, false, &objv_tracker));
+      if (retcode < 0) {
+        ldout(cct, 20) << "failed to read bucket shard status: "
+            << cpp_strerror(retcode) << dendl;
+        return set_cr_error(retcode);
+      }
+
+      // update bucket_status after a shard is done with current gen
+      bucket_status.shards_done_with_gen.push_back(true);
+
+      using WriteCR = RGWSimpleRadosWriteCR<rgw_bucket_sync_status>;
+      yield call(new WriteCR(sync_env->async_rados, sync_env->svc->sysobj,
+                        status_obj, bucket_status, &objv_tracker, false));
+      if (retcode < 0) {
```

Comment:
since this is a read-modify-write and we're using `objv_tracker` to detect racing writes, we need to handle the `ECANCELED` errors when it detects that another write happened in the meantime

that means all of this read-modify-write logic needs to be in a loop that keeps retrying until the write returns something other than `ECANCELED`